lib/stdlib/src/uri_string.erl

```diff
@@ -0,0 +1,2067 @@
+-spec is_hex_digit(char()) -> boolean().
+is_hex_digit(C)
+  when $0 =< C, C =< $9;$a =< C, C =< $f;$A =< C, C =< $F -> true;
+is_hex_digit(_) -> false.
+
+
+%% Remove enclosing brackets from binary
+-spec remove_brackets(binary()) -> binary().
+remove_brackets(<<$[/utf8, Rest/binary>>) ->
+    {H,T} = split_binary(Rest, byte_size(Rest) - 1),
+    case T =:= <<$]/utf8>> of
+        true -> H;
+        false -> Rest
+    end;
+remove_brackets(Addr) -> Addr.
+
+
+%%-------------------------------------------------------------------------
+%% Helper functions for calculating the parsed binary.
+%%-------------------------------------------------------------------------
+-spec calculate_parsed_scheme(binary(), binary()) -> binary().
+calculate_parsed_scheme(Input, <<>>) ->
+    strip_last_char(Input, [$:]);
+calculate_parsed_scheme(Input, Unparsed) ->
+    get_parsed_binary(Input, Unparsed).
+
+
+-spec calculate_parsed_part(binary(), binary()) -> binary().
+calculate_parsed_part(Input, <<>>) ->
+    strip_last_char(Input, [$?,$#]);
+calculate_parsed_part(Input, Unparsed) ->
+    get_parsed_binary(Input, Unparsed).
+
+
+-spec calculate_parsed_userinfo(binary(), binary()) -> binary().
+calculate_parsed_userinfo(Input, <<>>) ->
+    strip_last_char(Input, [$?,$#,$@]);
+calculate_parsed_userinfo(Input, Unparsed) ->
+    get_parsed_binary(Input, Unparsed).
+
+
+-spec calculate_parsed_host_port(binary(), binary()) -> binary().
+calculate_parsed_host_port(Input, <<>>) ->
+    strip_last_char(Input, [$:,$?,$#,$/]);
+calculate_parsed_host_port(Input, Unparsed) ->
+    get_parsed_binary(Input, Unparsed).
+
+
+calculate_parsed_query_fragment(Input, <<>>) ->
+    strip_last_char(Input, [$#]);
+calculate_parsed_query_fragment(Input, Unparsed) ->
+    get_parsed_binary(Input, Unparsed).
+
+
+get_port(<<>>) ->
+    undefined;
+get_port(B) ->
+    try binary_to_integer(B)
+    catch
+        error:badarg ->
+            throw({error, invalid_uri, B})
+    end.
+
+
+%% Strip last char if it is in list
+strip_last_char(<<>>, _) -> <<>>;
+strip_last_char(Input, [C0]) ->
+    case binary:last(Input) of
+        C0 ->
+            init_binary(Input);
+        _Else ->
+            Input
+    end;
+strip_last_char(Input, [C0,C1]) ->
+    case binary:last(Input) of
+        C0 ->
+            init_binary(Input);
+        C1 ->
+            init_binary(Input);
+        _Else ->
+            Input
+    end;
+strip_last_char(Input, [C0,C1,C2]) ->
+    case binary:last(Input) of
+        C0 ->
+            init_binary(Input);
+        C1 ->
+            init_binary(Input);
+        C2 ->
+            init_binary(Input);
+        _Else ->
+            Input
+    end;
+strip_last_char(Input, [C0,C1,C2,C3]) ->
+    case binary:last(Input) of
+        C0 ->
+            init_binary(Input);
+        C1 ->
+            init_binary(Input);
+        C2 ->
+            init_binary(Input);
+        C3 ->
+            init_binary(Input);
+        _Else ->
+            Input
+    end.
+
```

Comment:
Why do have so many clauses for different number of chars,  is it some kind of optimization?  Could you not do
` strip_last_char(Input, Chars) ->
    case lists:is_member(binary:last(Input), Chars) of
        true ->
             init_binary(Input); 
        false ->
            Chars
    end.   
`

Reply:
Yes, it is an optimization. I've measured uri_string:parse with this function and it got 10% slower, so this function is needed to outperform a list based parser implementation. :)

Reply:
Maybe write a comment about it in the code

Reply:
Done!